src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -118,15 +120,210 @@ void CoreWorkerDirectTaskSubmitter::AddWorkerLeaseClient(
     const SchedulingKey &scheduling_key) {
   client_cache_->GetOrConnect(addr.ToProto());
   int64_t expiration = current_time_ms() + lease_timeout_ms_;
-  LeaseEntry new_lease_entry = LeaseEntry(std::move(lease_client), expiration, 0,
-                                          assigned_resources, scheduling_key);
+  LeaseEntry new_lease_entry = LeaseEntry(std::move(lease_client), expiration, 0, false,
+                                          0, assigned_resources, scheduling_key);
   worker_to_lease_entry_.emplace(addr, new_lease_entry);
 
   auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
   RAY_CHECK(scheduling_key_entry.active_workers.emplace(addr).second);
   RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
 }
 
+void CoreWorkerDirectTaskSubmitter::ReturnWorker(const rpc::WorkerAddress addr,
+                                                 bool was_error,
+                                                 const SchedulingKey &scheduling_key) {
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+  RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
+  auto &lease_entry = worker_to_lease_entry_[addr];
+  RAY_CHECK(lease_entry.lease_client);
+  RAY_CHECK(lease_entry.tasks_in_flight == 0);
+  RAY_CHECK(lease_entry.WorkerIsStealing() == false);
+
+  // Decrement the number of active workers consuming tasks from the queue associated
+  // with the current scheduling_key
+  scheduling_key_entry.active_workers.erase(addr);
+  if (scheduling_key_entry.CanDelete()) {
+    // We can safely remove the entry keyed by scheduling_key from the
+    // scheduling_key_entries_ hashmap.
+    scheduling_key_entries_.erase(scheduling_key);
+  }
+
+  auto status =
+      lease_entry.lease_client->ReturnWorker(addr.port, addr.worker_id, was_error);
+  if (!status.ok()) {
+    RAY_LOG(ERROR) << "Error returning worker to raylet: " << status.ToString();
+  }
+  worker_to_lease_entry_.erase(addr);
+}
+
+bool CoreWorkerDirectTaskSubmitter::FindOptimalVictimForStealing(
+    const SchedulingKey &scheduling_key, rpc::WorkerAddress thief_addr,
+    rpc::Address *victim_raw_addr) {
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+
+  // Check that there is at least one worker (other than the thief) with the current
+  // SchedulingKey and that there are stealable tasks
+  if (max_tasks_in_flight_per_worker_ == 1 ||
+      scheduling_key_entry.active_workers.size() <= 1 ||
+      !scheduling_key_entry.StealableTasks()) {
+    return false;
+  }
+  RAY_CHECK(scheduling_key_entry.active_workers.size() > 1);
```

Comment:
These checks don't really do anything since they're just the opposite of the if checks above. Can you remove them?